cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -233,6 +233,124 @@ private class GuardConditionFromBinaryLogicalOperator extends GuardConditionImpl
   }
 }
 
+/**
+ * Holds if `ir` controls `block`, meaning that `block` is only
+ * entered if the value of this condition is `v`. This helper
+ * predicate does not necessarily hold for binary logical operations like
+ * `&&` and `||`. See the detailed explanation on predicate `controls`.
+ */
+private predicate controlsBlock(IRGuardCondition ir, BasicBlock controlled, AbstractValue v) {
+  exists(IRBlock irb |
+    ir.valueControls(irb, v) and
+    nonExcludedIRAndBasicBlock(irb, controlled) and
+    not isUnreachedBlock(irb)
+  )
+}
+
+private class GuardConditionFromNotExpr extends GuardConditionImpl {
+  IRGuardCondition ir;
+
+  GuardConditionFromNotExpr() {
+    // When `!` is applied to an integer (such as `x`) the generated IR looks
+    // like:
+    // ```
+    // r1(glval<int>) = VariableAddress[myInt] :
+    // r2(int)        = Load[x]                : &:r1, m1_6
+    // r3(int)        = Constant[0]            :
+    // r4(bool)       = CompareEQ              : r2, r3
+    // ```
+    // And so the `IRGuardCondition` for an expression such as `if(!x)` is the
+    // `CompareEQ` instruction. However, users often expect the `x` to also
+    // be a guard condition. But from the perspective of the IR the `x` is just
+    // the left-hand side of a comparison against 0 so it's not included as a
+    // normal `IRGuardCondition`. So to align with user expectations we make
+    // that `x` a `GuardCondition`.
```

Comment:
This is no longer up-to-date, I think, as you dropped the non-boolean restriction.

Reply:
Good point. Fixed in b39a932ae1c220f5f188fde039d1a1d7228b4314